hypothesis-python/src/hypothesis/strategies/_internal/strategies.py

```diff
@@ -564,14 +564,53 @@ def __repr__(self) -> str:
         )
 
     def calc_label(self) -> int:
-        return combine_labels(
-            self.class_label,
-            *(
-                (calc_label_from_hash(self.elements),)
-                if is_hashable(self.elements)
-                else ()
-            ),
-        )
+        # strategy.label is effectively an under-approximation of structural
+        # equality (i.e., some strategies may have the same label when they are not
+        # structurally identical). More importantly for calculating the
+        # SampledFromStrategy label, we might have hash(s1) != hash(s2) even
+        # when s1 and s2 are structurally identical. For instance:
+        #
+        #   s1 = st.sampled_from([st.none()])
+        #   s2 = st.sampled_from([st.none()])
+        #   assert hash(s1) != hash(s2)
+        #
+        # (see also test cases in test_labels.py).
+        #
+        # We therefore use the labels of any component strategies when calculating
+        # our label, and only use the hash if it is not a strategy.
+        #
+        # That's the ideal, anyway. In reality the logic is more complicated than
+        # necessary in order to be efficient in the presence of (very) large sequences:
+        # * add an unabashed special case for range, to avoid iteration over an
+        #   enormous range when we know it is entirely integers.
+        # * if there is at least one strategy in self.elements, use strategy label,
+        #   and the element hash otherwise.
+        # * if there are no strategies in self.elements, take the hash of the
+        #   entire sequence. This prevents worst-case performance of hashing each
+        #   element when a hash of the entire sequence would have sufficed.
+        #
+        # The worst case performance of this scheme is
+        # itertools.chain(range(2**100), [st.none()]), where it degrades to
+        # hashing every int in the range.
```

Comment:
OK, yeah, that would really bad if it was reachable.

Note though that you can't actually feed that in this far; `self.elements = cu.check_sample(elements, "sampled_from")` in the init method will try materializing elements into a tuple, and so I think you're limited to representable-in-memory numbers of elements.  That's still plenty of room to be slow, but not as slow as the explicitly-special-cased `range(...)`.

Reply:
Nice, forgot we materialized that. (though we do *not* materialize `range` in `check_sample`).